aten/src/ATen/native/cpu/SortingKernel.cpp

```diff
@@ -108,32 +113,75 @@ static void sort_kernel(
     ) {
       using scalar_t = typename std::remove_pointer<decltype(values)>::type;
+      using accscalar_t = typename AccType<scalar_t>::type;
       auto values_accessor = StridedRandomAccessor<scalar_t>(
         values, values_dim_stride);
       auto indices_accessor = StridedRandomAccessor<int64_t>(
         indices, indices_dim_stride);
-      auto composite_accessor = CompositeRandomAccessorCPU<
-        decltype(values_accessor), decltype(indices_accessor)
-      >(values_accessor, indices_accessor);
-
-      if (descending) {
-        if (stable) {
-          std::stable_sort(composite_accessor, composite_accessor + dim_size,
-            KeyValueCompDesc<scalar_t>());
+      // If the dimension of the sort axis is not physically dense 
+      // or the data type of values is BFloat16, 
+      // allocate a new array of accscalar_t data type and copy data in.
+      if(values_dim_stride != 1 || std::is_same<scalar_t, BFloat16>::value) {
+        accscalar_t *accvalue = new accscalar_t[dim_size];
+        int64_t *accvalue_indices = new int64_t[dim_size];
+        for (int64_t j = 0; j < dim_size; j++) {
+          accvalue[j] = values_accessor[j];
+          accvalue_indices[j] = indices_accessor[j];
         }
-        else {
-          std::sort(composite_accessor, composite_accessor + dim_size,
-            KeyValueCompDesc<scalar_t>());
+        auto accvalue_accessor = StridedRandomAccessor<accscalar_t>(
+          accvalue, 1);
+        auto accvalue_indices_accessor = StridedRandomAccessor<int64_t>(
+          accvalue_indices, 1);
+        auto composite_accessor = CompositeRandomAccessorCPU<
+          decltype(accvalue_accessor), decltype(accvalue_indices_accessor)
+        >(accvalue_accessor, accvalue_indices_accessor);
+        if (descending) {
+          if (stable) {
+            std::stable_sort(composite_accessor, composite_accessor + dim_size,
+              KeyValueCompDesc<accscalar_t>());
+          }
+          else {
+            std::sort(composite_accessor, composite_accessor + dim_size,
+              KeyValueCompDesc<accscalar_t>());
+          }
+        } else {
+          if (stable) {
+            std::stable_sort(composite_accessor, composite_accessor + dim_size,
+              KeyValueCompAsc<accscalar_t>());
+          }
+          else {
+            std::sort(composite_accessor, composite_accessor + dim_size,
+              KeyValueCompAsc<accscalar_t>());
+          }
         }
-      }
-      else {
-        if (stable) {
-          std::stable_sort(composite_accessor, composite_accessor + dim_size,
-            KeyValueCompAsc<scalar_t>());
+        for (int64_t j = 0; j < dim_size; j++) {
+          values_accessor[j] = accvalue[j];
+          indices_accessor[j] = accvalue_indices[j];
         }
-        else {
-          std::sort(composite_accessor, composite_accessor + dim_size,
-            KeyValueCompAsc<scalar_t>());
+        delete [] accvalue;
+        delete [] accvalue_indices;
```

Comment:
Again, you can probably use `std::copy` with accessors directly. Also, we can remove `new` and `delete`, and create empty tensors with `at::empty` and access their pointers with `data_ptr<scalar_t>` instead.